doc/oomd/SONiC OOM daemon.md

```diff
@@ -0,0 +1,267 @@
+  * [3.2 Default OOMD config](#32-default-oomd-config)
+  * [3.3 ConfigDB Schema](#34-configdb-schema)
+  * [3.4 CLI](#35-cli)
+- [4 Error handling](#error-handling)
+- [5 Serviceability and Debug](#serviceability-and-debug)
+- [6 Unit Test](#unit-test)
+- [8 References](#references)
+
+
+# About this Manual
+This document provides a detailed description on the new features for:
+ - Protect SONiC system memory by OOM daemon.
+
+ - Re-start docker container when container in OOM.
+
+ - OOM daemon config command.
+
+ - OOM daemon high level design.
+
+ - OOM daemon ConfigDB schema
+
+### Linux OOM
+ - OOM: Out Of Memory
+ - OOM Killer: OOM handler in Linux kernel
+     - Customize by /proc/sys/vm/panic_on_oom:
+         - 0: Kill process based on /proc/$PID/oom_score_adj
+         - 1: Kill current process who request for memory
+         - 2: Kernel panic
+     - Known issue:
+         - Can't protect key process:
+             - When panic_on_oom is 0, set oom_score_adj will partially mitigate this issue.
+         - Kernel panic will harm the system stability.
+ - How to trigger OOM:
+      - When system no available memory.
+      - SysRq: Send f to /proc/sysrq-trigger.
+      - Trigger by cgroup:
+           - cgroup can set memory limit.
+           - When cgroup memory usage reach limit, depends on memory.oom_control:
+                - 1: send OOM signal to system.
+                - 0: Hang current request memory process.
+      - Trigger by docker:
+           - Every docker container has a cgroup under: /sys/fs/cgroup/memory/docker/
+           - container can set cgroup policy by 'docker run' parameter.
+           - container cgroup can trigger OOM
+
+### SONiC memory issue solved by this feature.
+ - Currently SONiC enabled OOM killer, and set /proc/sys/vm/panic_on_oom to 2, which will trigger kernal panic when OOM. This is by design to protect SONiC key process and container.
+ - A typical switch device have 4 GB memory and sonic usually will use 1.5 GB for dockers, and 500 MB for system process. so there will be 2 GB free memory for user. 
+ - SONiC does not enable swap for most device. This means when OOM happen, system will not get memory with swap, this means OOM killer will be triggered very fast on SONiC.
+ - When user run some command trigger OOM, SONiC will kernel panic. for example:
+   - Multiple user login to device, some service may create 10+ concurrent sessions login to device.
+   - Some user script/command take too much memory, currently 'show' command will take 70 MB memory.
+ - To fix this issue, we need 2 features: login session limit and OOM daemon.
+
+# 1 Functional Requirement
+ - Monitor docker OOM status and re-start docker when OOM happen.
+     - Some docker container will set memory limit and disable trigger system OOM killer.
+     - These docker container will hang when OOM happen.
+     - OOMD will subscribe OOM event of these docker, and re-start container when container OOM happen.
```

Comment:
> OOMD will subscribe OOM event of these docker, and re-start container when container OOM happen [](http://example.com/codeflow?start=6&length=96)

Not a design question, FYI: @yozhao101 found that when container OOM happens and hang, `docker stats` will also hang. Need to find a new way to monitor the memory usage.

Reply:
@liuh-80 @qiluo-msft Another finding is if the OOM occurs in a container, restarting this container such as using `systemctl restart <service_name>.service` will also fail. 

Specifically the `systemctl` command will first execute the script to invoke the command `docker stop <service_name>` to stop the container and then invoke the `docker start <service_name>.service to start the container. But these operations will fail since all the docker CLI commands will hang for this container.

Reply:
I will try following solution which I think will slove this issue:
1. Find all related pid from the docker container cgroup in /sys/fs/cgroup/memory/docker/{container id}/tasks
2. Kill some procress  in this container by pid, this will free memory, then the docker container will not hang anymore.
3. restart this container.

I will test this and give update here. if this works OOMD can do this check when container OOM happen

Reply:
@liuh-80 Another finding is that we can retrieve the root process ID of container from the file `/sys/fs/cgroup/memory/docker/<container id>/cgroup.procs` and the first PID is the root process ID. After that, we can terminate that process to restart the container automatically.

The difference between these two files `cgroup.procs` and `tasks` is that if one process in this cgroup forks multiple threads, then the parent process ID will be put in the file `cgroup.procs`. Parent process ID and all its child thread IDs will be put in the file `tasks`.